src/annotations/mod.rs

```diff
@@ -0,0 +1,481 @@
+        #[derive(Serialize)]
+        struct ObjectInfo {
+            created_by: TaskId,
+            created_at: usize,
+            name: Option<String>,
+            kind: Option<String>,
+        }
+
+        /// Information about a task.
+        #[derive(Serialize)]
+        struct TaskInfo {
+            created_by: TaskId,
+            first_step: usize,
+            last_step: usize,
+            name: Option<String>,
+        }
+
+        #[derive(Debug, Serialize)]
+        enum AnnotationEvent {
+            SemaphoreCreated(ObjectId),
+            SemaphoreClosed(ObjectId),
+            SemaphoreAcquireFast(ObjectId, usize),
+            SemaphoreAcquireBlocked(ObjectId, usize),
+            SemaphoreAcquireUnblocked(ObjectId, TaskId, usize),
+            SemaphoreTryAcquire(ObjectId, usize, bool),
+            SemaphoreRelease(ObjectId, usize),
+
+            TaskCreated(TaskId, bool),
+            TaskTerminated,
+
+            Random,
+            Tick,
+        }
+
+        #[derive(Serialize)]
+        struct EventInfo(
+            // which task did something/yielded?
+            TaskId,
+            // backtrace
+            Option<Vec<Frame>>,
+            // event kind
+            AnnotationEvent,
+            // (if available,) clock of the task
+            // TODO: should always be available?
+            Option<VectorClock>,
+            // which other tasks were available to schedule, if this was a scheduled tick
+            Option<Vec<TaskId>>,
+        );
+
+        #[derive(Default, Serialize)]
+        struct AnnotationState {
+            version: usize,
+            files: Vec<FileInfo>,
+            #[serde(skip)]
+            path_to_file: HashMap<String, usize>,
+            functions: Vec<FunctionInfo>,
+            #[serde(skip)]
+            name_to_function: HashMap<String, usize>,
+            objects: Vec<ObjectInfo>,
+            tasks: Vec<TaskInfo>,
+            events: Vec<EventInfo>,
+
+            #[serde(skip)]
+            last_runnable_ids: Option<Vec<TaskId>>,
+            #[serde(skip)]
+            last_task_id: Option<TaskId>,
+            #[serde(skip)]
+            max_task_id: Option<TaskId>,
+        }
+
+        impl Serialize for VectorClock {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::ser::Serializer,
+            {
+                use serde::ser::SerializeSeq;
+                let mut seq = serializer.serialize_seq(Some(self.time.len()))?;
+                for e in &self.time {
+                    seq.serialize_element(e)?;
+                }
+                seq.end()
+            }
+        }
+
+        impl Serialize for TaskId {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::ser::Serializer,
+            {
+                usize::from(*self).serialize(serializer)
+            }
+        }
```

Comment:
Nit-ish:

Should all the serde stuff be moved somewhere else and put behind a feature flag "serde" (and "annotation" would enable "serde" as well)? It doesn't belong here and it also might be useful for other stuff to have a way to get serde for the types.